ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/target/VirtualArtifactRepository.java

```diff
@@ -0,0 +1,207 @@
+
+	public VirtualArtifactRepository(IProvisioningAgent agent, ITargetLocation targetLocation) {
+		super(agent, NAME + getLocationSafe(targetLocation), targetLocation.getType(), null,
+				URI.create(MEMORY_PREFIX + UUID.randomUUID()), DESCRIPTION + '\n' + targetLocation.serialize(), null,
+				null);
+		Assert.isTrue(targetLocation.isResolved());
+		for (TargetBundle targetBundle : targetLocation.getBundles()) {
+			if (!targetBundle.getStatus().isOK()) {
+				PDECore.log(Status.warning(NLS.bind(Messages.VirtualArtifactRepository_0, targetBundle)));
+				continue;
+			}
+			BundleInfo bundleInfo = targetBundle.getBundleInfo();
+			IArtifactKey artifactKey = BundlesAction.createBundleArtifactKey(bundleInfo.getSymbolicName(),
+					bundleInfo.getVersion());
+			IArtifactDescriptor artifactDesriptor = new ArtifactDescriptor(artifactKey);
+			artifacts.put(artifactDesriptor, bundleInfo);
+		}
+		for (TargetFeature targetFeature : targetLocation.getFeatures()) {
+			IArtifactKey artifactKey = FeaturesAction.createFeatureArtifactKey(targetFeature.getId(),
+					targetFeature.getVersion());
+			IArtifactDescriptor artifactDesriptor = new ArtifactDescriptor(artifactKey);
+			artifacts.put(artifactDesriptor, targetFeature.getFeatureModel());
+		}
+	}
+
+	private static String getLocationSafe(ITargetLocation targetLocation) {
+		try {
+			return targetLocation.getLocation(false);
+		} catch (CoreException e) {
+			return "<unknown>"; //$NON-NLS-1$
+		}
+	}
+
+	@Override
+	public IStatus getRawArtifact(IArtifactDescriptor descriptor, OutputStream destination, IProgressMonitor monitor) {
+		Object artifactModel = artifacts.get(descriptor);
+		if (artifactModel == null) {
+			return Status.error(NLS.bind(Messages.VirtualArtifactRepository_1, descriptor));
+		}
+		try {
+			transferArtifact(artifactModel, destination);
+			return Status.OK_STATUS;
+		} catch (Exception e) {
+			return Status.error(e.getLocalizedMessage(), e);
+		}
+	}
+
+	private void transferArtifact(Object artifactModel, OutputStream destination) throws Exception {
+		if (artifactModel instanceof BundleInfo bundleInfo) {
+			URI location = bundleInfo.getLocation();
+			if (location == null) {
+				throw new FileNotFoundException(bundleInfo.getSymbolicName());
+			}
+			try (InputStream is = location.toURL().openStream()) {
+				is.transferTo(destination);
+			}
+		} else if (artifactModel instanceof IFeatureModel featureModel) {
+			String installLocation = featureModel.getInstallLocation();
+			if (installLocation != null) {
+				File featureJar = new File(installLocation);
+				if (featureJar.isFile()) {
+					Files.copy(featureJar.toPath(), destination);
+					return;
+				}
+			}
+			IFeature feature = featureModel.getFeature();
+			// Generate in-memory feature jar (with only the feature.xml)
+			Manifest manifest = new Manifest();
+			manifest.getMainAttributes().put(Attributes.Name.MANIFEST_VERSION, "1.0"); //$NON-NLS-1$
```

Comment:
I wonder if PDE or P2 considers the MANIFEST.MF or a feature-jar at any time and therefore if this is necessary?

Reply:
PDE usually only ever creates a default Manifest for features, so probably not... Perhaps there's a requirement when the features are signed, but that's not something that applies here.
So I guess it can go?

Reply:
> PDE usually only ever creates a default Manifest for features, so probably not... Perhaps there's a requirement when the features are signed, but that's not something that applies here. So I guess it can go?

Signing is indeed the only use-case where a Manifest is necessary, I'm not aware of any other one.
So yes I think it can go.